[Lens] Add internal CRUD api routes

nickofthyme opened the pull request:
## Summary

This adds basic Lens CRUD api routes using the Content Management system.

| Operation | URI |
|--------|--------|
| Create | `POST api/lens/visualizations` |
| Get | `GET api/lens/visualizations/{id}` |
| Search | `GET api/lens/visualizations?query=test` |
| Update | `PUT api/lens/visualizations/{id}` |
| Delete | `DELETE api/lens/visualizations/{id}` |

### Changes to Lens Content Management

The custom `update` method uses `soClient.create` under the hood for reasons (i.e. #160116). However, doing this acts as an update or create method with the provided `id`. I changed this behavior so now any update where the id is not found will return a `404` error.

Closes #221941
Closes #221942 - OpenAPI docs auto generate from route schema

### Testing

You can testing this locally in kibana dev console like so...

```
GET kbn:/api/lens/visualizations/<id>?apiVersion=1
```

> The `apiVersion` query param is needed to test `internal` api routes.

## Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.

<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

nickofthyme committed:
feat: add internal lens api routes

nickofthyme commented on src/platform/plugins/shared/dashboard/server/api/register_routes.ts:
```diff
@@ -222,17 +222,15 @@ export function registerAPIRoutes({
-                max: 1000,
-              })
-            ),
+            perPage: schema.number({
+              meta: {
+                description:
+                  'The number of dashboards to display on each page (max 1000). Default is "20".',
+              },
+              defaultValue: 20,
+              min: 1,
+              max: 1000,
+            }),
```
Using a `defaultValue` is enough to treat this as optional, but when also paired with `schema.maybe` the validated value will be `undefined` instead of `20`.

nickofthyme replied:
Mostly for @nickpeihl

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/api/routes/visualizations/create.ts:
```diff
@@ -0,0 +1,95 @@
+  { contentManagement }
+) => {
+  const createRoute = router.post({
+    path: `${PUBLIC_API_PATH}/visualizations`,
```
I used this as `lens/visualizations/` instead of `lens/` to be more explicit about the resource being change and in case there are other resources to add later.

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/api/routes/visualizations/create.ts:
```diff
@@ -0,0 +1,95 @@
+      validate: {
+        request: {
+          params: schema.object({
+            // should we allow custom ids?
+            id: schema.maybe(
+              schema.string({
+                meta: {
+                  description: 'The saved object ID of the Lens visualization.',
+                },
+              })
+            ),
```
Dashboards API allows setting a custom `id`, should we do the same?

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/api/routes/visualizations/create.ts:
```diff
@@ -0,0 +1,95 @@
+        .getForRequest({ request: req, requestHandlerContext: ctx })
+        .for(CONTENT_ID, PUBLIC_API_CONTENT_MANAGEMENT_VERSION);
+      let result;
+      try {
+        ({ result } = await client.create(data, options));
+      } catch (e) {
+        // TODO prevent duplicate titles?
+
+        if (e.isBoom && e.output.statusCode === 403) {
+          return res.forbidden();
+        }
```
Do we want to prohibit duplicate titles at the API level?

dej611 replied:
Good question.
@nickpeihl do you have an existing logic for that in presentation? I would like consistency on the API side with the UI, hence keep the title validation as well, but let's sync with the entire API ecosystem as I think it has priority over UI.
@ppisljar yes we do. It is not possible to save a visualisation with an existing title.

nickofthyme replied:
Interesting, I never realized that was the flow. I figured the save was disabled.

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/api/routes/visualizations/search.ts:
```diff
@@ -0,0 +1,113 @@
+        return res.badRequest();
+      }
+
+      const body = {
+        items: result.hits,
+        total: result.pagination.total,
+      };
+      return res.ok({ body });
```
Do we want to be consistent about return values for search APIs?

Also what is the need for returning the `total`, most public APIs do not include this info.

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/content_management/lens_storage.ts:
```diff
@@ -136,6 +136,9 @@ export class LensStorage extends SOContentStorage<LensCrudTypes> {
     // Save data in DB
     const soClient = await savedObjectClientFromRequest(ctx);
 
+    // since we use create below this is to throw if SO id not found
+    await soClient.get(CONTENT_ID, id);
```
Without this addition, the update call will create the SO even if it does not already exist.

nickofthyme commented on x-pack/platform/plugins/shared/lens/server/content_management/v1/cm_services.ts:
```diff
@@ -54,7 +54,7 @@ const lensSavedObjectSchema = schema.object(
 );
 
-const getResultSchema = schema.object(
+export const lensGetResultSchema = schema.object(
   {
     item: lensSavedObjectSchema,
     meta: schema.object(
```
Do we want to return all this other `meta` data from the API? Typical with get APIs, I'd expect just the value, not a nested object value (i.e. like [GitHub's get issue](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#get-an-issue)).

nickofthyme reviewed (commented):
A few thoughts to consider.

nickofthyme committed:
chore: update api specs and error handling

nickofthyme committed:
test: add basic lens api integration tests

nickofthyme committed:
Merge branch 'main' into new-lens-api

nickofthyme committed:
chore: fix test title

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/yarn_deduplicate'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/309696)
* Commit: baef217487a2f00f140eb8495b9ef18a3103f7c9

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/309696#019780b2-cc6a-4ff6-aac6-1c02629af5f4)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309696","number":309696,"commit":"baef217487a2f00f140eb8495b9ef18a3103f7c9"}],"number":309696}
buildkite-pr-comment-->

markov00 committed:
update codeowners file

dmlemeshko commented on src/platform/test/api_integration/apis/lens/visualizations/create/index.ts:
```diff
@@ -0,0 +1,30 @@
+export default function ({ getService, loadTestFile }: FtrProviderContext) {
+  const kibanaServer = getService('kibanaServer');
+  describe('visualizations - create', () => {
+    before(async () => {
+      await kibanaServer.importExport.load(
+        'src/platform/test/api_integration/fixtures/kbn_archiver/saved_objects/basic.json'
+      );
+    });
+
+    after(async () => {
+      await kibanaServer.savedObjects.cleanStandardList();
+      await kibanaServer.importExport.unload(
+        'src/platform/test/api_integration/fixtures/kbn_archiver/saved_objects/basic.json'
+      );
+    });
```
Nice to see archives loaded only one time for suite ❤️ Thank you!

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/309759)
* Commit: fb73fc25f282f76f6345fa850edd4fd034471ffd

### Failed CI Steps
* [Jest Tests #16](https://buildkite.com/elastic/kibana-pull-request/builds/309759#0197824b-acce-4f37-9466-bc0525ba9b95)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/309759#0197824b-acce-4f37-9466-bc0525ba9b95) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/309759/jobs/0197824b-acce-4f37-9466-bc0525ba9b95/artifacts/01978272-ef09-4824-8d6e-c28f6e1e399c) Jest Tests #<span></span>16 / RedirectWithDefaultEnvironment eventually renders the child element

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/f0829b7c1179ef998380153ad7dd5b6a3bee6ea2) | [after](https://github.com/elastic/kibana/commit/fb73fc25f282f76f6345fa850edd4fd034471ffd) | diff |
| --- | --- | --- | --- |
| `lens` | 561 | 563 | +2 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/f0829b7c1179ef998380153ad7dd5b6a3bee6ea2) | [after](https://github.com/elastic/kibana/commit/fb73fc25f282f76f6345fa850edd4fd034471ffd) | diff |
| --- | --- | --- | --- |
| `lens` | 660 | 662 | +2 |


</details>


### History
* :broken_heart: [Build #309696](https://buildkite.com/elastic/kibana-pull-request/builds/309696) failed baef217487a2f00f140eb8495b9ef18a3103f7c9

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309759","number":309759,"commit":"fb73fc25f282f76f6345fa850edd4fd034471ffd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309696","number":309696,"commit":"baef217487a2f00f140eb8495b9ef18a3103f7c9"}],"number":309759}
buildkite-pr-comment-->

ppisljar reviewed (approved):
nice,  LGTM!

afharo commented on src/platform/test/api_integration/fixtures/kbn_archiver/saved_objects/lens.json:
```diff
@@ -0,0 +1,315 @@
+{
```
Can we change the ownership of this file? I doubt that Core must be the owner. Any reasons for us to own this file?

nickofthyme replied:
Done ✅

nickofthyme committed:
chore: assign vis team as fixture CO

nickofthyme committed:
Merge branch 'main' into new-lens-api

nickofthyme committed:
Merge remote-tracking branch 'origin/new-lens-api' into new-lens-api

nickofthyme committed:
chore: fix bad CO merge

nickofthyme committed:
chore: fix bad CO merge

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/311808)
* Commit: 38a5438b58ba3e70c34a8b6bcc10ccc24afd8c6c

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/311808#01979f70-e362-423b-bcec-d04a639eb385)

### History
* :yellow_heart: [Build #309759](https://buildkite.com/elastic/kibana-pull-request/builds/309759) was flaky fb73fc25f282f76f6345fa850edd4fd034471ffd
* :broken_heart: [Build #309696](https://buildkite.com/elastic/kibana-pull-request/builds/309696) failed baef217487a2f00f140eb8495b9ef18a3103f7c9

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311808","number":311808,"commit":"38a5438b58ba3e70c34a8b6bcc10ccc24afd8c6c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309759","number":309759,"commit":"fb73fc25f282f76f6345fa850edd4fd034471ffd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309696","number":309696,"commit":"baef217487a2f00f140eb8495b9ef18a3103f7c9"}],"number":311808}
buildkite-pr-comment-->

nickofthyme committed:
chore: move lens api tests under x-pack

- fix paths
- update files after move
- move fixtures
- add lens test config
- update CO file with move

nickofthyme committed:
chore: update COs again

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/yarn_deduplicate'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/styled_components_mapping'

jbudz reviewed (approved):
`.buildkite/ftr_platform_stateful_configs.yml` LGTM

nickofthyme committed:
Merge branch 'main' into new-lens-api

nickofthyme committed:
chore: remove co presentation review changes

nickofthyme committed:
Merge remote-tracking branch 'origin/new-lens-api' into new-lens-api

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/312778)
* Commit: 9fdc36ba82cee6b527d2d7080e5a7b8988b26cf3

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/cae3861f5d4fe1fea0419eedbf852b460e3033ee) | [after](https://github.com/elastic/kibana/commit/9fdc36ba82cee6b527d2d7080e5a7b8988b26cf3) | diff |
| --- | --- | --- | --- |
| `lens` | 561 | 563 | +2 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/cae3861f5d4fe1fea0419eedbf852b460e3033ee) | [after](https://github.com/elastic/kibana/commit/9fdc36ba82cee6b527d2d7080e5a7b8988b26cf3) | diff |
| --- | --- | --- | --- |
| `lens` | 660 | 662 | +2 |


</details>


### History
* :yellow_heart: [Build #311819](https://buildkite.com/elastic/kibana-pull-request/builds/311819) was flaky 2cac74cc750ad195853441607867847f093d9be1
* :broken_heart: [Build #311808](https://buildkite.com/elastic/kibana-pull-request/builds/311808) failed 38a5438b58ba3e70c34a8b6bcc10ccc24afd8c6c
* :yellow_heart: [Build #309759](https://buildkite.com/elastic/kibana-pull-request/builds/309759) was flaky fb73fc25f282f76f6345fa850edd4fd034471ffd
* :broken_heart: [Build #309696](https://buildkite.com/elastic/kibana-pull-request/builds/309696) failed baef217487a2f00f140eb8495b9ef18a3103f7c9

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312778","number":312778,"commit":"9fdc36ba82cee6b527d2d7080e5a7b8988b26cf3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311819","number":311819,"commit":"2cac74cc750ad195853441607867847f093d9be1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311808","number":311808,"commit":"38a5438b58ba3e70c34a8b6bcc10ccc24afd8c6c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309759","number":309759,"commit":"fb73fc25f282f76f6345fa850edd4fd034471ffd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309696","number":309696,"commit":"baef217487a2f00f140eb8495b9ef18a3103f7c9"}],"number":312778}
buildkite-pr-comment-->

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.19:<br> - [[ska] relocate x-pack/test/cases_api_integration (#225056)](https://github.com/elastic/kibana/pull/225056)<br> - [[ska] [xpack] relocate platform tests (#225223)](https://github.com/elastic/kibana/pull/225223)<br> - [Move apps/security and apps/spaces tests under platform](https://github.com/elastic/kibana/pull/224464)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 223296
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

nickofthyme commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/225463">](https://github.com/elastic/kibana/pull/225463)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.1 -->

kibanamachine commented:
Looks like this PR has a backport PR but it still hasn't been merged. Please merge it ASAP to keep the branches relatively in sync.
cc: @nickofthyme

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 223296` locally
cc: @nickofthyme
